src/pubsub.c

```diff
@@ -615,6 +615,116 @@ void publishCommand(client *c) {
+        addReplyBulkLongLong(c, client->id);
+        addReplyBulkCString(c, "addr");
+        addReplyBulkCString(c, getClientPeerId(client));
+    }
+}
+
+/* Parse the arguments for PUBSUB SUBSCRIBERS command */
+static int parseArgsPubSubSubscribersCmd(client *c, pubsubSubscribersCmdArgs *args) {
+    if (c->argc < 3) {
+        addReplySubcommandSyntaxError(c);
+        return C_ERR;
+    }
+
+    /* Mandatory parameter - subscription_type */
+    char *type = c->argv[2]->ptr;
+    if (!strcasecmp(type, "channel")) {
+        args->subscription_dict = server.pubsub_channels;
+    } else if (!strcasecmp(type, "pattern")) {
+        args->subscription_dict = server.pubsub_patterns;
+    } else if (!strcasecmp(type, "shard-channel")) {
+        args->subscription_dict = server.pubsubshard_channels;
+    } else {
+        addReplySubcommandSyntaxError(c);
+        return C_ERR;
+    }
+
+    /* Optional parameter - match / count */
+    for (int i = 3; i < c->argc; i++) {
+        int moreargs = c->argc-i-1;
+        char *o = c->argv[i]->ptr;
+        if (!strcasecmp(o, "match") && moreargs) {
+            i++;
+            args->pat = c->argv[i]->ptr;
+            args->use_pattern = !(sdslen(args->pat) == 1 && args->pat[0] == '*');
+        } else if (!strcasecmp(o, "count") && moreargs) {
+            i++;
+            if (getRangeLongFromObjectOrReply(c, c->argv[i], -1, LONG_MAX, &args->count,
+                                             "count should be greater than or equal to -1") != C_OK) {
+                return C_ERR;
+            }
+            /* count arg provided as -1 signifies all the channels needs to be sent. */
+            if (args->count == -1) {
+                args->count = dictSize(args->subscription_dict);
+            }
+        } else {
+            addReplySubcommandSyntaxError(c);
+            return C_ERR;
+        }
+    }
+    return C_OK;
+}
+
+static void pubsubReplySubscribers(client *c) {
+    /* Initialize default args and parse if any. */
+    pubsubSubscribersCmdArgs args;
+    pubsubSubscribersCmdArgsInitializeDefault(&args);
+    if (parseArgsPubSubSubscribersCmd(c, &args) != C_OK) {
+        return;
+    }
+
+    void *dl = addReplyDeferredLen(c);
+    long long cnt = 0;
+
+    dictIterator *di = dictGetIterator(args.subscription_dict);
+    dictEntry *de;
+    while((de = dictNext(di)) != NULL && cnt < args.count) {
+        /* Filter the channels/subscriptions. */
+        sds channel = ((robj *)dictGetKey(de))->ptr;
+        if (args.use_pattern && !stringmatchlen(args.pat, sdslen(args.pat), channel, sdslen(channel), 0)) {
+            continue;
+        }
+        list *clients = dictGetVal(de);
+        serverAssert(clients != NULL);
+        addReplyBulkCString(c, channel);
+        addReplyMapLen(c, 1);
+        addReplyBulkCString(c, "subscribed-clients");
+        addReplyArrayLen(c, listLength(clients));
+        pubsubReplySubscribersClientInfo(c, clients);
+        cnt++;
```

Comment:
```suggestion
        cnt+=listLength(clients);
```
I would expect count to be number of subscriptions returned, not number of channels returned. I am okay returning more than 1000 subscriptions in order to return all the subscriptions from a client. If each channel has 10k clients, that is 10M responses, which seems like a high default.

Reply:
Nice suggestion. This should keep the results more bounded but we could breach the count in some cases 👍